apps/devhub/tasks.py

```diff
@@ -32,65 +32,116 @@
-        raise
+        # FIXME: Have validator return native Python datastructure.
+        return json.loads(result)
 
-    try:
-        # Try annotating the result with the "passed auto validation" status.
-        validation = json.loads(result)
-        # The automatic validation (which will automatically review and sign
-        # the file) is for unlisted (non-sideload) addons and beta versions.
-        # It only passes this automatic validation if there's no message with a
-        # signing_severity above trivial (so if there's any low/medium/high
-        # severity, it fails).
-        # Any "regular" messages from the amo-validator that should fail the
-        # auto validation now have a signing_severity level attached.
-        passed_auto_validation = False
-        if 'signing_summary' in validation:
-            summary = validation['signing_summary']
-            passed_auto_validation = (summary['low'] == 0 and
-                                      summary['medium'] == 0 and
-                                      summary['high'] == 0)
-        validation['passed_auto_validation'] = passed_auto_validation
-        result = json.dumps(validation)
-    except Exception:  # A bare `except:` would ignore system exceptions.
-        # Failed loading the result? No need to set passed_auto_validation.
-        pass
-
-    # Skip the "addon signed" warning if we're not signing.
-    if not settings.SIGNING_SERVER:
-        result = skip_signing_warning(result)
-    upload.validation = result
+
+@task(soft_time_limit=settings.VALIDATOR_TIMEOUT)
+def validate_file(file_id, **kw):
+    """Validate a File instance. If cached validation results exist, return
+    those, otherwise run the validator."""
+
+    if settings.VALIDATE_ADDONS:
+        file = File.objects.get(pk=file_id)
+        try:
+            return file.validation.validation
+        except FileValidation.DoesNotExist:
+            result = run_validator(file.current_file_path,
+                                   listed=file.version.addon.is_listed)
+            # FIXME: Have validator return native Python datastructure.
+            return json.loads(result)
+
+
+@task
+@write
+def handle_upload_validation_result(results, upload_id):
+    """Annotates a set of validation results and saves them to the given
+    FileUpload instance."""
+    result = annotate_validation_results(results)
+
+    upload = FileUpload.objects.get(pk=upload_id)
+    upload.validation = json.dumps(result)
     upload.save()  # We want to hit the custom save().
+    return upload
 
 
-def skip_signing_warning(result_json):
+@task
+@write
+def handle_file_validation_result(results, file_id):
+    """Annotates a set of validation results and saves them to the given
+    File instance."""
+    result = annotate_validation_results(results)
+
+    file_ = File.objects.get(pk=file_id)
+    return FileValidation.from_json(file_, result)
+
+
+@task
+def annotate_validation_results(results):
```

Comment:
I'm not really a huge fan of it either. I originally wrote it to accept a list, which would have one or two arguments, depending on whether the task group had two validation tasks or one. If it had two, the function wouldn't have different behavior, but would just add more annotations based on the additional data.

But the way I wrote the task dispatch code on the first go, it wound up just not being a task group, and therefore not being a list, so I decided to just go with that.

I'd really prefer it to be a function that just took two parameters, `results, previous_results=None`, but that would mean registering a stub task just to call this function, which seemed kind of messy.